### PR TITLE
Working on reducing latency in HMD

### PR DESCRIPTION
**Pull request by jherico**
This adds an explicit wait function in the presentation thread to allow as much time as possible after the last v-sync for the rendering thread / GPU to complete the next frame.  This should remove up to 1 frame of latency from our render->screen time.  The new behavior requires a plugin deriving from OpenGLDisplayPlugin to override the `prePresentWait` method and wait an appropriate amount of time there.  What's appropriate is entirely up to the plugin to determine.  For the Oculus implementation I am taking the approach of the total expected frame time minus 3 milliseconds.

* Exposes the ability to switch between the old and new behavior in the `Developer` > `Display` menu
* Exposes the Oculus performance HUD via the `Developer` > `Oculus` menu

When using the original behavior, the performance HUD reports a total latency of about 40-45 ms of latency (determined using the hardware latency tester integrated into the headset).

When using the new behavior, the HUD reports about 20-25 ms of latency.  For comparison, the Oculus World Demo reports a latency of about 18 ms, so we assume that's close to the minimum.  

This is a first pass that only includes the work for the Oculus plugin, but it can be tested on that platform.  Will do the Vive in a separate commit.

### CODE REVIEW

File: plugins/oculus/src/OculusDisplayPlugin.cpp [hunk 1]
```diff
@@ -24,10 +37,6 @@ void OculusDisplayPlugin::customizeContext() {
     _sceneLayer.ColorTexture[0] = _sceneFbo->color;
     // not needed since the structure was zeroed on init, but explicit
     _sceneLayer.ColorTexture[1] = nullptr;
-
-    enableVsync(false);
```
Comment (ZappoMan) [inline plugins/oculus/src/OculusDisplayPlugin.cpp]: is this just cruft? or was this causing problems?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 03df4f9b63a552bdcffa53558e79477f74aac56e available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by PhilipRosedale** [comment]
Will test on Vive

**Comment by PhilipRosedale** [comment]
Testing generates problems for me on Oculus DK2, windows, Oculus SDK 0.8: 

go to playa, float near stacks of RV's with desert in background.   Run dice.js so you see the three overlay icons.  Comparing release and this PR, turn head slowly back and forth ("no") and watch Icons.  With this PR, they jitter very frequently - several times a second - by a frame or so.  With the current release (4403) they are solid.

File: plugins/oculus/src/OculusDisplayPlugin.cpp (hunk 1 above)
Comment (jherico) [inline plugins/oculus/src/OculusDisplayPlugin.cpp]: it was cruft.  the vsync handling and mono / preview settings are all handled in the common HMD base class.

**Comment by jherico** [comment]
@PhilipRosedale  are they solid on this build if you disable the wait (in developer -> display, uncheck 'pre-present wait')

**Comment by hifi-gustavo** [comment]
Build console output for commit: 44a439e0cfbbbcba5dff6cb01d533ccc19734b37 available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4fa3e9caad31ad1a95955f5b58ef63db5b4be6e0 available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by samcake** [comment]
@jherico I m not sure you need the 2 glFLush in the escrow code, do you?

**Comment by samcake** [comment]
Sorry!

**Comment by jherico** [comment]
@samcake re: glFlush....  I was surprised, but until I added that the fence did not work at all.  See the contexts section of this page:  https://www.opengl.org/wiki/Sync_Object

**Comment by samcake** [comment]
ok

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4fa3e9caad31ad1a95955f5b58ef63db5b4be6e0 available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by jherico** [comment]
https://app.asana.com/0/27650181942747/98694020434656

**Comment by ZappoMan** [comment]
I think I'm seeing what Philip was seeing with this PR. Specifically in cellscience, I see a lot of "shaking" in the scene. It's like the cells are all twitching out. I don't see it in playa, or demo, but cellscience is really bad. It doesn't change when I enable or disable the "pre-present wait".

**Comment by ZappoMan** [comment]
Also - the VR menus are borked for the new items you added... not sure if it's a size limit issue or if it's something else.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0a062e9f397984e6aedadb1f4a1f08ff8866a2cd available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by PhilipRosedale** [comment]
@jherico have you worked on the Vive yet?  Ready to re-test?

**Comment by PhilipRosedale** [comment]
QA Failed:  On Oculus, latest build of this PR, there is very substantial/worse judder on looking at the table in demo and turning the head slowly from side to side, as compared to in release.   The objects on the table appear to jitter against the background/table.   

The setting of the developer->display->pre-present wait does not make any difference in this effect that I can see.

File: libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp [hunk 2]
```diff
@@ -508,6 +516,10 @@ void OpenGLDisplayPlugin::internalPresent() {
 void OpenGLDisplayPlugin::present() {
     incrementPresentCount();
     updateTextures();
+    if (_enableWait) {
+        prePresentWait();
+        updateTextures();
```
Comment (ZappoMan) [inline libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp]: Wait... so it looks like you're calling `updateTextures()` twice... is that what you want?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1c82f365b7b6d87c80f623db04cd0ddeb3d04698 available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3655d8d13e452cbc0b96e65cd8b57cbac2e6439a available at: https://gist.github.com/2c93c6db549604498664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7324.dmg

**Comment by PhilipRosedale** [comment]
@jherico Do you need to fix merge conflicts here and I can re-test?

**Comment by ZappoMan** [comment]
Closing this for now, until we've had a chance to discuss the design so that it doesn't have negative impact on low FPS situation.

**Pull request closed without merging**